util/topgen.py

```diff
@@ -663,32 +663,62 @@ def generate_top_ral(top: Dict[str, object],
     # Collect up the memories to add
     mems = []
     for item in list(top.get("memory", [])):
-        byte_write = ('byte_write' in item and
-                      item["byte_write"].lower() == "true")
-        data_intg_passthru = ('data_intg_passthru' in item and
-                              item["data_intg_passthru"].lower() == "true")
-        size_in_bytes = int(item['size'], 0)
-        num_regs = size_in_bytes // addrsep
-        swaccess = access.SWAccess('top-level memory',
-                                   item.get('swaccess', 'rw'))
-
-        mems.append(window.Window(name=item['name'],
-                                  desc='(generated from top-level)',
-                                  unusual=False,
-                                  byte_write=byte_write,
-                                  data_intg_passthru=data_intg_passthru,
-                                  validbits=regwidth,
-                                  items=num_regs,
-                                  size_in_bytes=size_in_bytes,
-                                  offset=int(item["base_addr"], 0),
-                                  swaccess=swaccess))
+        mems.append(create_mem(item, item['name'], addrsep, regwidth))
+
+    # Top-level may override the mem setting. Store the new type to name_to_block
+    # If no other instance uses the orignal type, delete it
+    list_orginal_types = []
+    for module in top['module']:
+        if 'memory' in module.keys() and len(module['memory']) > 0:
+            block = deepcopy(name_to_block[module['type']])
+            name_to_block[module['name']] = block
+            inst_to_block[module['name']] = module['name']
```

Comment:
This is cute! But I don't think we have any assertion that says module names and types are distinct, so I think this could see collisions.

Could we coin a new "type name"? Maybe something like `newtype = '{}_{}'.format(module['type'], idx)` where `idx` is a counter for the loop? That should always be unique, but we should probably also add an assertion to check we didn't mess up: `assert newtype not in name_to_block`.

Reply:
good point. I changed to `newtype = '{}_{}'.format(module['type'], module['name'])`, idx is less meaningful. Thanks

Reply:
That's much nicer! Thanks.